### ISSUE TITLE
Tool scripts: Add a paragraph on how Inputs are handled in a tool script and how to get node references

### ISSUE THREAD

**Issue by golddotasksquestions** [godotengine/godot-docs#5262]
**Your Godot version:** Godot 3.3.3 stable

**Issue description:**
Over the past three years I have wasted literally weeks on trying to figure out how to get my tool scripts to respond to Input. Eventually I have given up. 

The[ "Running code in the Editor" ](https://docs.godotengine.org/en/stable/tutorials/misc/running_code_in_the_editor.html)docs page imho should mention, how in tool scripts, neither `_input(event)`, `_unhandled_input(event)` , `_gui_input(event)`, nor `Input.is_action_pressed("action")` works.

If should also mention the fact that 
`Input.is_mouse_button_pressed(BUTTON_MIDDLE) `
`Input.is_key_pressed(KEY_SPACE)`
`Input.is_joy_button_pressed(0,JOY_BUTTON_0)`
_**do**_ work when used inside `_process()`

Imho it also really needs to mention how getting node references in `_ready()` and `onready var my_node = $MyNode` does not work in tool scripts but getting those refererences in _init() does work:
```

var my_node = null
func _init():
	my_node = $MyNode
```

If these things would have been on this page, it would have saved so much time and frustration and would have made using tool scripts so much more accessible and more useful to me.

**URL to the documentation page (if already existing):**
https://docs.godotengine.org/en/stable/tutorials/misc/running_code_in_the_editor.html

**Comment by pycbouh** [comment]
> The[ "Running code in the Editor" ](https://docs.godotengine.org/en/stable/tutorials/misc/running_code_in_the_editor.html)docs page imho should mention, how in tool scripts, neither `_input(event)`, `_unhandled_input(event)` , `_gui_input(event)`

I don't think this has anything to do with tool scripts. They obviously can handle all of those events, as evident from the plugins that do so. The reason why some interactions don't work, though, is that the elements of your scene are put into their own viewport, which is blocked by the editor itself for editor tools to work.

> nor `Input.is_action_pressed("action")` works

This is also supported by tool scripts, in theory, but in practice the editor context probably doesn't know your project's action map and therefore doesn't detect any actions.

But whatever the reason, this can of course be documented if there is any confusion. But one should take care not to make it seem like tool scripts are somehow magically incapable of those things by their nature.

> it also really needs to mention how getting node references in `_ready()` and `onready var my_node = $MyNode` does not work in tool scripts but getting those refererences in `_init()` **_does_** work

That doesn't make sense at all, however. For one, if it works on init, it will work on ready. But it will never work on init in the first place, because that's a constructor and the children themselves have not been initialized and the node produced by the script has not been added to the tree. Maybe there are some shenanigans with adding a tool script to a loaded scene without reloading it, but it's probably just a temporary bug due to hot-loading and should not be documented as the way it supposedly works.

**Comment by hsandt** [comment]
Actually I was wondering if I needed an early return in input methods, such as:

```gdscript
func _unhandled_input(event: InputEvent):
    if Engine.is_editor_hint():
        return
```

and apparently it's not the case, so I still find it useful if the `@tool` script page could list the exact methods called by `@tool` script in the editor. From testing, the methods that run in editor in `@tool` scripts are:

- `_ready`
- `_enter_tree`
- `_exit_tree`
- `_process`
- `_physics_process`
- `_draw`
- property setters

Am I missing anything?

Currently, the doc only mentions `_process` and the property setters. It doesn't even mention `_draw` although the 3 examples at the beginning are about drawing debug shapes!

So it would make sense to list all those methods, and for extra clarity indicate that the input methods are not handled (but could be handled via `_process` as indicated in OP).

(edited as there were many more methods called than I thought)

**Comment by YuriSizov** [comment]
@hsandt There is no explicit list of methods that run under `@tool`, because script in its entirety runs under `@tool`. Input is a bit of an anomaly in that regard:

- Your edited scene is placed under a sub-viewport within the editor, and sub-viewports [do not receive input events on their own](https://docs.godotengine.org/en/4.6/tutorials/inputs/inputevent.html#how-does-it-work) in Godot.
- A sub-viewport would receive input when put under a GUI `SubViewportContainer` node which also receives input, and in our case the sub-viewport for the edited scene is inside of a sub-viewport container. However, our sub-viewport also [explicitly turns `disable_input` ON](https://github.com/godotengine/godot/blob/8421c61c7edd9a19e3f147a7a57813db23539ed5/editor/editor_node.cpp#L8632-L8638), meaning it will ignore input even if it reaches the sub-viewport.
- A sub-viewport container runs the `_input` processing [only when directly focused](https://github.com/godotengine/godot/blob/8421c61c7edd9a19e3f147a7a57813db23539ed5/scene/gui/subviewport_container.cpp#L149-L159) (which never happens in the editor, I believe), and otherwise runs only the `_unhandled_input` processing (which, as the name suggests, only happens when the event has not been handled by something else, severely limiting the chances of that to happen).
- On top of that, a sub-viewport container is explicitly [not passing through input events in the editor](https://github.com/godotengine/godot/blob/8421c61c7edd9a19e3f147a7a57813db23539ed5/scene/gui/subviewport_container.cpp#L181-L187). And also any viewport that is inside of your edited scene is [prevented from handling input in the editor](https://github.com/godotengine/godot/blob/8421c61c7edd9a19e3f147a7a57813db23539ed5/scene/main/viewport.cpp#L3473-L3475).

Scripts themselves work just fine, though. If you turn OFF `disable_input` on the sub-viewport containing the edited scene and then `push_input(event)` to it, that event will reach the script no problem.

That is to say that should this information be reflected in the docs, it's important to not misinform the reader with implications that there are some limitations to the scripts themselves. Tool scripts are used for plugins, and handle input events just fine, just like you'd expect them to. It's just that the editor, understandably, tries to prevent your scene from receiving input.
